rest/oidc_api_test.go

```diff
@@ -1,139 +1,865 @@
+		})
+	}
+}
+
+// Checks End to end OIDC Authorization Code flow.
+func TestOpenIDConnectAuth(t *testing.T) {
+	mockAuthServer := newMockAuthServer()
+	mockAuthServer.Start()
+	defer mockAuthServer.Shutdown()
+
+	issuerFoo := mockAuthServer.URL + "/foo"
+	mockAuthServer.SetOption("issuer", issuerFoo)
+
+	type test struct {
+		name                      string
+		providers                 auth.OIDCProviderMap
+		defaultProvider           string
+		authURL                   string
+		register                  bool
+		includeAccessToken        bool
+		notConfiguredProvider     bool
+		wantUsername              string
+		wantNoCode                bool
+		wantCallbackError         bool
+		wantUntoldProvider        bool
+		wantDiscoveryError        bool
+		wantCallbackTokenExgError bool
+		wantRefreshTokenExgError  bool
+		wantCallbackNoIDToken     bool
+		wantRefreshNoIDToken      bool
+	}
+	tests := []test{
+		{
+			name: "new user auto registration enabled single provider",
+			providers: auth.OIDCProviderMap{
+				"foo": &auth.OIDCProvider{
+					Name:               "foo",
+					Issuer:             issuerFoo,
+					ClientID:           "baz",
+					UserPrefix:         "foo",
+					ValidationKey:      base.StringPointer("qux"),
+					Register:           true,
+					DiscoveryURI:       issuerFoo + auth.OIDCDiscoveryConfigPath,
+					IncludeAccessToken: true,
+				},
+			},
+			defaultProvider:    "foo",
+			wantUsername:       "foo_noah",
+			authURL:            "/db/_oidc?provider=foo&offline=true",
+			register:           true,
+			includeAccessToken: true,
+		}, {
+			name: "new user auto registration disabled single provider",
+			providers: auth.OIDCProviderMap{
+				"foo": &auth.OIDCProvider{
+					Name:               "foo",
+					Issuer:             issuerFoo,
+					ClientID:           "baz",
+					UserPrefix:         "foo",
+					ValidationKey:      base.StringPointer("qux"),
+					Register:           false,
+					DiscoveryURI:       issuerFoo + auth.OIDCDiscoveryConfigPath,
+					IncludeAccessToken: true,
+				},
+			},
+			defaultProvider:    "foo",
+			wantUsername:       "foo_noah",
+			authURL:            "/db/_oidc?provider=foo&offline=true",
+			register:           false,
+			includeAccessToken: true,
+		}, {
+			name: "new user auto registration enabled no access token single provider",
+			providers: auth.OIDCProviderMap{
+				"foo": &auth.OIDCProvider{
+					Name:               "foo",
+					Issuer:             issuerFoo,
+					ClientID:           "baz",
+					UserPrefix:         "foo",
+					ValidationKey:      base.StringPointer("qux"),
+					Register:           true,
+					DiscoveryURI:       issuerFoo + auth.OIDCDiscoveryConfigPath,
+					IncludeAccessToken: true,
+				},
+			},
+			defaultProvider:    "foo",
+			wantUsername:       "foo_noah",
+			authURL:            "/db/_oidc?provider=foo&offline=true",
+			register:           true,
+			includeAccessToken: false,
```

Comment:
Why doesn't this match the value on the provider?  More generally, is this necessary as a parameter, or can it be taken from the provider?